src/hub/dataload/sources/drugcentral/drugcentral_dump.py

```diff
@@ -1,15 +1,134 @@
+import csv
 import os
 
 import biothings
+import psycopg2
+
 import config
 
 biothings.config_for_app(config)
 
-from config import DATA_ARCHIVE_ROOT
-from biothings.hub.dataload.dumper import DummyDumper
+from biothings.hub.dataload.dumper import BaseDumper
 
+from config import DATA_ARCHIVE_ROOT, DRUGCENTRAL_PASSWORD
 
-class DrugCentralDumper(DummyDumper):
 
+class DrugCentralDumper(BaseDumper):
     SRC_NAME = "drugcentral"
     SRC_ROOT_FOLDER = os.path.join(DATA_ARCHIVE_ROOT, SRC_NAME)
+
+    # More info about the public DrugCentral Postgres database https://drugcentral.org/download
+    HOST = "unmtid-dbs.net"
+    PORT = 5433
+    DATABASE = "drugcentral"
+    USER = "drugman"
+
+    def prepare_client(self):
+        # Connect to the database
+        self.client = psycopg2.connect(
+            database=self.DATABASE,
+            user=self.USER,
+            password=DRUGCENTRAL_PASSWORD,
+            host=self.HOST,
+            port=self.PORT,
+        )
+        # Create a cursor
+        self._cursor = self.client.cursor()
+
+    @property
+    def cursor(self):
+        # Create a cursor if it doesn't exist
+        if not getattr(self, "_cursor", None):
+            if self.client:
+                self._cursor = self.client.cursor()
+            else:
+                self.logger.debug(
+                    "_cursor was not initialized and client is not active."
+                )
+                self._cursor = None
+        return self._cursor
+
+    def unprepare(self):
+        # Set the state of the cursor to None
+        self._cursor = None
+        return super().unprepare()
+
+    def release_client(self):
+        # Disconnect from the database
+        if self.client:
+            self.client.close()
+            self.client = None
+
+    def download(self, remotefile, localfile):
+        # Download the data from the database and write it to a CSV file
+
+        # Create the local folders if they don't exist
+        self.prepare_local_folders(localfile)
+
+        table_name, columns = remotefile.get("table_name"), remotefile.get("columns")
+
+        cursor = self.cursor
+        cursor.execute(f"SELECT {columns} FROM {table_name}")
+
+        column_names = [desc[0] for desc in cursor.description]
+
+        # Open the CSV file and write the column names
+        with open(os.path.join(localfile), "w", newline="") as file:
+            writer = csv.writer(file)
+            writer.writerow(column_names)
+
+            # Write the rows one by one
+            row = cursor.fetchone()
+            while row:
+                writer.writerow(row)
+                row = cursor.fetchone()
```

Comment:
Have you checked out this `itersize` setting on `cursor`:

https://www.psycopg.org/docs/cursor.html#cursor.itersize

You may put this as a class property like `CURSOR_ITERSIZE`, default can be 100 to be safe, then you can just loop through `cursor` here:

```
for row in cursor:
    writer.writenow(row)
```

Reply:
Here we need to do a few things, first define the itersize variable and then set it to the cursor when we prepare_client. 
```
class DrugCentralDumper(BaseDumper):
    SRC_NAME = "drugcentral"
    SRC_ROOT_FOLDER = os.path.join(DATA_ARCHIVE_ROOT, SRC_NAME)

    # The number of rows to fetch at a time with the cursor
    CURSOR_ITERSIZE = 100
    ...

    def prepare_client(self):
        ....
        # Create a cursor
        self._cursor = self.client.cursor()
        # Set the number of rows to fetch at a time
        self._cursor.itersize = self.CURSOR_ITERSIZE
```
And finally loop over the cursor:
```
        with open(os.path.join(localfile), "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(column_names)

            row_count = 0

            # Loop directly over the cursor
            for row in cursor:
                writer.writerow(row)
                row_count += 1

        self.logger.debug(
            f"Retrieved {row_count} rows from table: {table_name} and saved to {localfile}"
        )
```
in commit f48f51b